Update tests for relaxed <select> parser

josepharhar opened the pull request:
This PR updates the tree-construction dat files for the HTML change which will allow additional tags within `<select>`:
https://github.com/whatwg/html/pull/10557

josepharhar committed:
Update tests for relaxed <select> parser

This PR updates the tree-construction dat files for the HTML change
which will allow additional tags within <select>:
https://github.com/whatwg/html/pull/10557

josepharhar commented:
I'm not sure what the best practice is for rebaselining errors, but for now I removed all errors from affected tests. There are probably errors in tests I didn't change which may need to be rebaselined as well.

annevk commented:
Will there be a separate PR for new tests?

flavorjones commented:
FWIW when CI workflows are enabled, Nokogiri (downstream) tests will fail. I've started working on a branch with the proposed changes from https://github.com/whatwg/html/pull/10557

josepharhar committed:
add new tests

josepharhar commented:
> Will there be a separate PR for new tests?

I added new test cases to webkit02 including:
- Reconstructing formatting contexts inside and outside of select: https://github.com/whatwg/html/pull/10557#discussion_r1796173325
- Test cases I added a while ago currently in select-parsing.tentative.html: https://github.com/web-platform-tests/wpt/blob/7275b2c6bf4b9312a1945a451a5333e8922ee524/html/semantics/forms/the-select-element/customizable-select/select-parsing.tentative.html

josepharhar committed:
add input in select test case

flavorjones commented:
Nokogiri work-in-progress at https://github.com/sparklemotion/nokogiri/pull/3317

flavorjones commented:
@josepharhar I've got a question about two tests that are very similar. Zooming in on this one from `tree-construction/tests1.dat`:

```
#data
<select><b><option><select><option></b></select>
#errors
#document
| <html>
|   <head>
|   <body>
|     <select>
|       <b>
|         <option>
|     <b>
|     <select>
|       <b>
|         <option>
```

Nokogiri is constructing a different tree:

```
<body>
  <select>
    <b>
      <option>
  <b>
    <select>
      <option>
```

and I wanted to ask for a double-check that the test's assertion is correct, before I dive into Nokogiri's parser. Thank you!

josepharhar commented:
Thanks for asking!

It looks like Nokogiri is nesting one `<select>` inside the other, and my spec/chromium change intentionally fully closes the outer select before inserting the new one, so yeah I think the test's assertion is correct.

Here's the relevant part of the spec PR: https://whatpr.org/html/10557/parsing.html#:~:text=A%20start%20tag%20whose%20tag%20name%20is%20%22select%22

As for the `<b>` in between the `<select>` and the `<option>`, I think there is a mismatch between my spec PR and the chromium implementation - specificially the chromium implementations stops in this case before reconstructing formatting contexts, which isn't in the spec: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/parser/html_tree_builder.cc;l=951;drc=9776ce6636ab1d68d2b1c4f1f719eb71becf39a3

I'll take a closer look at which we should do and get back to you on that. Thanks!

flavorjones commented:
> It looks like Nokogiri is nesting one <select> inside the other

No, unless I'm misunderstanding you (totally possible), I don't think that's right. Here's the tree from my earlier comment displayed a bit more graphically:

```
body
└── select
    ├── b
    │   └── option
    b
    └── select
        └── option
```

The selects are not nested. Just wanted to clarify.

flavorjones commented:
Thanks for replying so quickly.

> It looks like Nokogiri is nesting one `select` inside the other

No, sorry, unless I'm misunderstanding your comment, this is not a correct description of what Nokogiri's parser is doing. Here's a more graphical representation of the tree from my previous comment:


```
body
├── select
│   └── b
│       └── option
└── b
    └── select
        └── option
```

The select tags are not nested. Just wanted to clarify.

josepharhar commented:
Ah whoops, I failed to read the tree properly 😅 

I looked into why chromium is putting the b inside the select instead of the other way around, and i found that the `</b>` tag in the test is what's making the difference - without it I am getting the same output as Nokogiri.

This is happening due to this call to the adoption agency algorithm when `</b>` is parsed: https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inbody:adoption-agency-algorithm-4

I verified that chromium does the same thing as Nokogiri when I comment out that call to the adoption agency algorithm.

There's a lot of steps in that algorithm and I'm having a hard time wrapping my head around it, but does Nokogiri have that call to the adoption agency algorithm too? Or perhaps the implementations of the algorithm are different?

flavorjones commented:
@josepharhar Thanks again for your kind reply!

Yes, Nokogiri's libgumbo has implemented the adoption agency algorithm, and I have confirmed that in these tests we are invoking it. It's helpful to know this is likely the source of the behavior difference, so I'll focus my efforts on making sure it matches the current spec (though I'll note it passes every other test in this suite ... 🤷).

flavorjones commented:
OK, I think I know what's going on here. I think Chromium has missed this change:

![image](https://github.com/user-attachments/assets/3df6901e-e264-4532-a4f5-15552c114fd8)

https://github.com/whatwg/html/pull/10557/files#diff-41cf6794ba4200b839c53531555f0f3998df4cbb01a4d5cb0b94e3ca5e23947dR124618

If `select` is not in the list for default scope tags, then this adoption agency step:

> If formattingElement is in the [stack of open elements](https://whatpr.org/html/10557/parsing.html#stack-of-open-elements), but the element is not [in scope](https://whatpr.org/html/10557/parsing.html#has-an-element-in-scope), then this is a [parse error](https://whatpr.org/html/10557/parsing.html#parse-errors); return.

will not trigger a parse-error-and-return, and the algorithm will continue. But `select` _is_ is the list now, and so the formatting element `b` is _not_ in scope.

If I remove the `select` tag from the default scope tags in Nokogiri's libgumbo, then behavior matches the test (and presumably chromium).

Can you check to see if my hunch is right?

josepharhar committed:
Add select scope

josepharhar commented:
Thanks so much for figuring this out! Yeah I totally missed that in the chromium implementation but I just added it and updated the tests here.

flavorjones commented:
@josepharhar That's great! Thank you!

I've got a patch to get the error messages to a point where libgumbo is passing, would you mind taking a look and potentially applying to this PR? (Renamed to `.txt` to get by the github attachment filter.)

[select-test-errors.patch.txt](https://github.com/user-attachments/files/17498078/select-test-errors.patch.txt)

josepharhar committed:
Fixes to the proposed tests at html5lib/html5lib-tests#178

josepharhar commented:
Looks good, thanks! I applied it.

Sometimes I wonder why chromium [throws away parse errors](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/parser/html_tree_builder.cc;l=3164;drc=d815f515138991af2aa5b1d07c64906fd8a7366b).

untitaker commented on tree-construction/webkit02.dat:
```diff
@@ -533,14 +534,13 @@ eof-in-math
 1:1: ERROR: Expected a doctype token
-1:46: ERROR: Premature end of file. Currently open tags: html, body, table, tbody, tr, td, select.
-#document
+1:38: ERROR: Premature end of file. Currently open tags: html, body, table, tbody, tr, td, select, optgroup.#document
```
```suggestion
1:38: ERROR: Premature end of file. Currently open tags: html, body, table, tbody, tr, td, select, optgroup.
#document
```

flavorjones replied:
👍 Thank you for spotting this.

untitaker commented:
the changes to `#errors` make `webkit02` highly inconsistent and break existing norms in e.g. parse5. It's recommended to align with https://github.com/whatwg/html/issues/1339

josepharhar committed:
Fix #document at end of line

Co-authored-by: Markus Unterwaditzer <markus-github@unterwaditzer.net>

zcorpan commented on tree-construction/webkit01.dat:
```diff
@@ -437,33 +437,41 @@ no-doctype
-|       <select>
-|         <option>
-|           "G"
+|     <select>
+|       <option>
+|         "B"
+|     <select>
+|       <option>
+|         "C"
+|     <select>
+|       <option>
+|         "D"
+|     <select>
+|       <option>
+|         "E"
+|     <select>
+|       <option>
+|         "F"
+|     <select>
+|       <option>
+|         "G"
+|     <select>
```
This test expectation should be reverted per https://github.com/whatwg/html/pull/10557#discussion_r1829039074

josepharhar replied:
done

zcorpan commented on tree-construction/webkit02.dat:
```diff
@@ -438,34 +439,34 @@ eof-in-math
 |       <optgroup>
 |         <option>
-|       <hr>
+|         <hr>
```
This should be reverted per https://github.com/whatwg/html/pull/10557#discussion_r1828988927

josepharhar replied:
done

zcorpan commented on tree-construction/webkit02.dat:
```diff
@@ -551,4 +552,140 @@ eof-in-math
+|     <select>
+|       <div>
+|         <i>
+|       <option>
+|         <i>
```
`<i>` should be the parent per https://github.com/whatwg/html/pull/10557#discussion_r1828992609

josepharhar replied:
done

josepharhar committed:
Add changes from spec PR

josepharhar commented:
I just updated this PR, which mostly involved reverting stuff. Should all be up to date and in sync with the whatwg/html PR now.

flavorjones commented:
@josepharhar Twelve of the tests changed or added in the recent commits do not have the correct number of error messages in the test data.

I'm not sure if you wish to fix those up given the @untitaker feedback above, but if so here are the failures from the Nokogiri test suite. Please let me know if I can help format these for you.

<details><summary>nokogiri libgumbo error match failures</summary>

```
Finished in 1.434869s, 4618.5409 runs/s, 79923.6865 assertions/s.

  1) Failure:
TestHtml5TreeConstructionTests2#test_37__script_on [test/html5/test_tree_construction.rb:131]:
Expected 2 errors for <!DOCTYPE html><select><optgroup><option></optgroup><option><select><option>, found:
1:61: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, select, option.
<!DOCTYPE html><select><optgroup><option></optgroup><option><select><option>
                                                            ^.
Expected: 2
  Actual: 1

  2) Failure:
TestHtml5TreeConstructionTests2#test_37__script_off [test/html5/test_tree_construction.rb:131]:
Expected 2 errors for <!DOCTYPE html><select><optgroup><option></optgroup><option><select><option>, found:
1:61: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, select, option.
<!DOCTYPE html><select><optgroup><option></optgroup><option><select><option>
                                                            ^.
Expected: 2
  Actual: 1

  3) Failure:
TestHtml5TreeConstructionWebkit01#test_31__script_on [test/html5/test_tree_construction.rb:131]:
Expected 9 errors for <select><option>A<select><option>B<select><option>C<select><option>D<select><option>E<select><option>F<select><option>G<select>, found:
1:1: ERROR: Expected a doctype token
<select><option>A<select><option>B<select><option>C<select><option>D<select><option>E<select><option>F<select><option>G<select>
^
1:18: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, select, option.
<select><option>A<select><option>B<select><option>C<select><option>D<select><option>E<select><option>F<select><option>G<select>
                 ^
1:52: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, option, select, option.
<select><option>A<select><option>B<select><option>C<select><option>D<select><option>E<select><option>F<select><option>G<select>
                                                   ^
1:86: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, option, select, option.
<select><option>A<select><option>B<select><option>C<select><option>D<select><option>E<select><option>F<select><option>G<select>
                                                                                     ^
1:120: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, option, select, option.
<select><option>A<select><option>B<select><option>C<select><option>D<select><option>E<select><option>F<select><option>G<select>
                                                                                                                       ^.
Expected: 9
  Actual: 5

  4) Failure:
TestHtml5TreeConstructionWebkit01#test_31__script_off [test/html5/test_tree_construction.rb:131]:
Expected 9 errors for <select><option>A<select><option>B<select><option>C<select><option>D<select><option>E<select><option>F<select><option>G<select>, found:
1:1: ERROR: Expected a doctype token
<select><option>A<select><option>B<select><option>C<select><option>D<select><option>E<select><option>F<select><option>G<select>
^
1:18: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, select, option.
<select><option>A<select><option>B<select><option>C<select><option>D<select><option>E<select><option>F<select><option>G<select>
                 ^
1:52: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, option, select, option.
<select><option>A<select><option>B<select><option>C<select><option>D<select><option>E<select><option>F<select><option>G<select>
                                                   ^
1:86: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, option, select, option.
<select><option>A<select><option>B<select><option>C<select><option>D<select><option>E<select><option>F<select><option>G<select>
                                                                                     ^
1:120: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, option, select, option.
<select><option>A<select><option>B<select><option>C<select><option>D<select><option>E<select><option>F<select><option>G<select>
                                                                                                                       ^.
Expected: 9
  Actual: 5

  5) Failure:
TestHtml5TreeConstructionWebkit02#test_40__script_on [test/html5/test_tree_construction.rb:131]:
Expected 4 errors for <select><button><select></select></button></select>, found:
1:1: ERROR: Expected a doctype token
<select><button><select></select></button></select>
^
1:17: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, select, button.
<select><button><select></select></button></select>
                ^
1:25: ERROR: End tag 'select' isn't allowed here. Currently open tags: html, body.
<select><button><select></select></button></select>
                        ^
1:34: ERROR: End tag 'button' isn't allowed here. Currently open tags: html, body.
<select><button><select></select></button></select>
                                 ^
1:43: ERROR: End tag 'select' isn't allowed here. Currently open tags: html, body.
<select><button><select></select></button></select>
                                          ^.
Expected: 4
  Actual: 5

  6) Failure:
TestHtml5TreeConstructionWebkit02#test_38__script_off [test/html5/test_tree_construction.rb:131]:
Expected 3 errors for <select><button>button</select>, found:
1:1: ERROR: Expected a doctype token
<select><button>button</select>
^
1:23: ERROR: End tag 'select' isn't allowed here. Currently open tags: html, body, select, button.
<select><button>button</select>
                      ^.
Expected: 3
  Actual: 2

  7) Failure:
TestHtml5TreeConstructionWebkit02#test_41__script_off [test/html5/test_tree_construction.rb:131]:
Expected 2 errors for <select><button><div><select></select>, found:
1:1: ERROR: Expected a doctype token
<select><button><div><select></select>
^
1:22: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, select, button, div.
<select><button><div><select></select>
                     ^
1:30: ERROR: End tag 'select' isn't allowed here. Currently open tags: html, body.
<select><button><div><select></select>
                             ^.
Expected: 2
  Actual: 3

  8) Failure:
TestHtml5TreeConstructionWebkit02#test_41__script_on [test/html5/test_tree_construction.rb:131]:
Expected 2 errors for <select><button><div><select></select>, found:
1:1: ERROR: Expected a doctype token
<select><button><div><select></select>
^
1:22: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, select, button, div.
<select><button><div><select></select>
                     ^
1:30: ERROR: End tag 'select' isn't allowed here. Currently open tags: html, body.
<select><button><div><select></select>
                             ^.
Expected: 2
  Actual: 3

  9) Failure:
TestHtml5TreeConstructionWebkit02#test_38__script_on [test/html5/test_tree_construction.rb:131]:
Expected 3 errors for <select><button>button</select>, found:
1:1: ERROR: Expected a doctype token
<select><button>button</select>
^
1:23: ERROR: End tag 'select' isn't allowed here. Currently open tags: html, body, select, button.
<select><button>button</select>
                      ^.
Expected: 3
  Actual: 2

 10) Failure:
TestHtml5TreeConstructionWebkit02#test_40__script_off [test/html5/test_tree_construction.rb:131]:
Expected 4 errors for <select><button><select></select></button></select>, found:
1:1: ERROR: Expected a doctype token
<select><button><select></select></button></select>
^
1:17: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, select, button.
<select><button><select></select></button></select>
                ^
1:25: ERROR: End tag 'select' isn't allowed here. Currently open tags: html, body.
<select><button><select></select></button></select>
                        ^
1:34: ERROR: End tag 'button' isn't allowed here. Currently open tags: html, body.
<select><button><select></select></button></select>
                                 ^
1:43: ERROR: End tag 'select' isn't allowed here. Currently open tags: html, body.
<select><button><select></select></button></select>
                                          ^.
Expected: 4
  Actual: 5

 11) Failure:
TestHtml5TreeConstructionTests7#test_17__script_off [test/html5/test_tree_construction.rb:131]:
Expected 2 errors for <!doctype html><select><select>X, found:
1:24: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, select.
<!doctype html><select><select>X
                       ^.
Expected: 2
  Actual: 1

 12) Failure:
TestHtml5TreeConstructionTests7#test_17__script_on [test/html5/test_tree_construction.rb:131]:
Expected 2 errors for <!doctype html><select><select>X, found:
1:24: ERROR: Start tag 'select' isn't allowed here. Currently open tags: html, body, select.
<!doctype html><select><select>X
                       ^.
Expected: 2
  Actual: 1
```

</summary>

josepharhar commented:
> Please let me know if I can help format these for you.

That would be great! I could try but I might mess it up 😅

flavorjones commented:
@josepharhar I've attached a patch file that indicates the number and nature of the error messages generated by libgumbo.

I want to acknowledge @untitaker's comment above that these error messages do not follow the format of other existing errors or the format suggested by https://github.com/whatwg/html/issues/1339, but this is the best I can do in the time I've allotted.

josepharhar committed:
Update error messages

josepharhar commented:
Thanks for generating those errors! I applied them in the latest commit

josepharhar committed:
Make <input> close <select>

sirreal commented on tree-construction/tests7.dat:
```diff
@@ -200,20 +200,20 @@ X</listing>
 |     <select>
-|     <input>
-|     "X"
+|       <input>
+|       "X"
```
Is this change correct? I'm looking at the updated standard and I understand that `<input>` should pop the `<select>` off the stack. If Chrome has implemented the change, that's the behavior I'm observing. [Live DOM Viewer link](https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C!doctype%20html%3E%3Cselect%3E%3Cinput%3EX).

<kbd><img width="698" height="648" alt="dom-viewer" src="https://github.com/user-attachments/assets/466e1511-2c5f-4931-8e69-b4015559e3c4" /></kbd>


Here's the relevant section of the current version of the standard for parsing in body (emphasis mine):

> A start tag whose tag name is "input"
> If the parser was created as part of the [HTML fragment parsing algorithm](https://html.spec.whatwg.org/multipage/parsing.html#html-fragment-parsing-algorithm) …
> 
> If the [stack of open elements](https://html.spec.whatwg.org/multipage/parsing.html#stack-of-open-elements) [has a select element in scope](https://html.spec.whatwg.org/multipage/parsing.html#has-an-element-in-scope):
> - [Parse error](https://html.spec.whatwg.org/multipage/parsing.html#parse-errors).
> - **Pop elements from the [stack of open elements](https://html.spec.whatwg.org/multipage/parsing.html#stack-of-open-elements) until a [select](https://html.spec.whatwg.org/multipage/form-elements.html#the-select-element) element has been popped from the stack.**

```suggestion
|     <input>
|     "X"
```

sirreal replied:
This looks the same as https://github.com/html5lib/html5lib-tests/pull/178/commits/7a7be4520dd5074c6c0c6bfc30bdb1b4ac2a7bd0. I suspect this case was overlooked and the input should close the select as I suggested here.

josepharhar replied:
Yes I overlooked that, it is now fixed. Thanks!

josepharhar committed:
make <input> close <select> again

josepharhar committed:
add selectedcontent tests

josepharhar commented:
Thanks @sirreal, I added the tests you provided to webkit02.dat. They pass in chromium.

sirreal reviewed (approved):
This is great. I'm external to this project, but have been using it and this helped adapt the WordPress HTML API to the customizable `<select>` changes.

For folks who aren't following, https://github.com/whatwg/html/pull/10548 has landed so the changes are now live in the [HTML Standard](https://html.spec.whatwg.org/).

gsnedders committed:
Empty commit to force CI to re-run

gsnedders commented:
> I think the errors are still a huge departure from existing style and will break a few testsuites. but I don't have commit rights on this repo so take my opinion with a grain of salt.

None of the CI failures seem to be because of the format, so I don't think this is an actual concern?

gsnedders reviewed (approved):
I've been somewhat out of touch of the parser for a while, but given it's landed in HTML, and Chromium implements this, and html5gum and skeleton pass the tests here, this seems fine.

@josepharhar do you have any suggestion for a commit message to squash this under?

untitaker commented:
@gsnedders to be fair I don't remember the details but I don't think the ci coverage is comprehensive. but for the sake of keeping up w the standard I think we can also deal with this once somebody complains.

annevk commented:
@josepharhar will you get these imported into WPT? Thanks!

josepharhar commented:
Here is a WPT PR: https://github.com/web-platform-tests/wpt/pull/54888
